### PR TITLE
Compressor Optimizer

### PR DESCRIPTION
**Pull request by Tabaie**
This PR implements issue https://github.com/Consensys/linea-monorepo/issues/366.

The compressor now attempts to recompress everything from scratch upon encountering a full blob, before attempting to bypass compression altogether.

### Checklist

* [ ] I wrote new tests for my new core changes.
* [x] I have successfully run tests, style checker and build against my new changes locally.
* [x] I have informed the team of any breaking changes if there are any.

### CODE REVIEW

**Comment by gbotrel** [comment]
So to sum up, this is a non-breaking change, we could compress better (by ~7%) if we re-compress the full blob each time we attempt to append a block; but doing so is too slow so you propose to keep the original method, that would give a "preliminary result", and change the result behind the scenes between calls?

Wouldn't this have some side effects @jpnovais ? (i.e basically the compressor could say "we compressed block 1, it takes N Bytes, current blob is now at 100kB", then recompress it with more context and update the internal state to "current blob is 98kB" without notifying the coordinator. ). 

Re implementation, before introducing an async optimizer, I'ld prefer to understand perf constraints better; i.e. how long it takes now, how long it would take if we recompress all the blob at each append, and within what limit we need operate . i.e. if we say the compressor could take as much as XXXms , then we may just want to have a simpler cleaner code and kill this async pattern.

**Comment by jpnovais** [comment]
My input on this optimization is:

Context:
- The coordinator does not actively keep track of the current blob size, that is the responsibility of the compressor, which is aware of blob limit. 
- Coordonator calls `fun CanWrite(data: ByteArray, data_len: Int): Boolean` before calling Write, if returns false then coordinator knows the blob is full and starts a new one

My take based on the above:
- As mentioned, Asyc can be problematic so I would avoid it. Also, I think we don't need it here.
- My suggestion: when coordinator calls `CanWrite` before returning false, the compressor would try to perform the **full compression** an see if that extra blob fits. 
  - **pros**: 
    -  lazy computation - it only does the "full re-compression" when it reaches the limit;
    - 100% transparent to the coordinator and keeps same API. This leaves room for further internal code performance optimizations ;

@Tabaie @gbotrel WDYT?

**Comment by gbotrel** [comment]
Yep that would make less CPU use to do it only when full . But this last call to "CanWrite" may be 10x slower than the previous calls.
So what are the bounds perf-wise we need to operate in? If a call to CanWrite takes 500ms is that acceptable?

**Comment by jpnovais** [comment]
> So what are the bounds perf-wise we need to operate in? If a call to CanWrite takes 500ms is that acceptable? (not saying it will, just want a rough order of magnitude of bound)

It's ok to have a call to CanWrite that takes 500ms **at the of the blob**, as long as the preceding calls are not affected timewise.

**Comment by Tabaie** [comment]
I agree, doing it synchronously at the end is a good idea. In fact it's similar to the "no compress" logic we already have.

**Review by gbotrel — DISMISSED** [review]
LGTM 👍 I would probably just add one or two test to ensure this is correctly triggered and that the internal state is correctly reset after

**Comment by Tabaie** [comment]
> > The difference, from the user's point of view, is output nondeterminism, as it will differ based on how much time the compressor has had to optimize.
> 
> What does this mean, "_based on how much time the compressor has had to optimize_" from the coordinator's PoV?

This was for the parallel optimizer so it no longer applies. Removing from description.

**Comment by Tabaie** [comment]
> > > The difference, from the user's point of view, is output nondeterminism, as it will differ based on how much time the compressor has had to optimize.
> > 
> > 
> > What does this mean, "_based on how much time the compressor has had to optimize_" from the coordinator's PoV?
> 
> This was for the parallel optimizer so it no longer applies. Removing from description.

@jpnovais Does it look good now?

**Review by jpnovais — APPROVED** [review]

**Comment by jpnovais** [comment]
@Tabaie sure. Sorry for the delay and not reverting on the 1st comment.

File: prover/lib/compressor/blob/v1/blob_maker.go [hunk 1]
```diff
@@ -410,3 +443,10 @@ func (bm *BlobMaker) RawCompressedSize(data []byte) (int, error) {
 
 	return n, nil
 }
+
+func wrapError(err error, format string, args ...any) error {
```
Comment (ivokub) [inline prover/lib/compressor/blob/v1/blob_maker.go:447]: Seems the method is only used once. I would remove it for now.

File: prover/lib/compressor/blob/v1/blob_maker.go [hunk 2]
```diff
@@ -191,45 +189,80 @@ func (bm *BlobMaker) Write(rlpBlock []byte, forceReset bool) (ok bool, err error
...
+	recompressionAttempted := false
+	revert := func() error { // from this point on, we may have recompressed the entire payload in one go
+		// that makes the compressor's own Revert method unusable.
+		bm.header.removeLastBlock()
+		if !recompressionAttempted { // fast path for most "CanWrite" calls
+			return bm.compressor.Revert()
+		}
+		// we can't use the compressor's own Revert method because we tried to compress in one go.
+		bm.compressor.Reset()
+		_, err := bm.compressor.Write(payload[:prevLen])
```
Comment (ivokub) [inline prover/lib/compressor/blob/v1/blob_maker.go:210]: Cannot we have issues here? payload refers to the internal buf of compressor and we write it back to the compressor. Couldn't we potentially mess up the buffer due to overwrites?

File: prover/lib/compressor/blob/v1/blob_maker.go [hunk 3]
```diff
@@ -191,45 +189,80 @@ func (bm *BlobMaker) Write(rlpBlock []byte, forceReset bool) (ok bool, err error
...
 	// check that the header + the compressed data fits in the blob
-	fitsInBlob := encode.PackAlignSize(bm.buf.Len()+bm.compressor.Len(), fr381.Bits-1) <= bm.Limit
-	if !fitsInBlob {
-		// first thing to check is if we bypass compression, would that fit?
+	if !fitsInBlob() {
+		recompressionAttempted = true
+
+		// first thing to check is whether we can fit the block if we recompress everything in one go, known to achieve a higher ratio.
+		bm.compressor.Reset()
+		if _, err = bm.compressor.Write(payload); err != nil {
```
Comment (ivokub) [inline prover/lib/compressor/blob/v1/blob_maker.go:220]: Similarly here, we write back the internal buf of compressor to compressor. I wonder if we may have overwrites here.

File: prover/lib/compressor/blob/v1/blob_maker.go [hunk 4]
```diff
@@ -191,45 +189,80 @@ func (bm *BlobMaker) Write(rlpBlock []byte, forceReset bool) (ok bool, err error
...
+			err = fmt.Errorf("when recompressing the blob: %w", err)
+
+			if innerErr := revert(); innerErr != nil {
+				err = fmt.Errorf("%w\n\tto recover from write failure: %w", innerErr, err)
+			}
+
+			return false, err
```
Comment (ivokub) [inline prover/lib/compressor/blob/v1/blob_maker.go:227]: Could simplify error logic using [`errors.Join`](https://pkg.go.dev/errors#Join). In this case we could do:
```go
innerErr := revert()
return false, errors.Join(err, innerErr)
```
as `errors.Join` discards nil errors.

File: prover/lib/compressor/blob/v1/blob_maker.go [hunk 5]
```diff
@@ -169,17 +168,16 @@ func (bm *BlobMaker) Write(rlpBlock []byte, forceReset bool) (ok bool, err error
 		if innerErr := bm.compressor.Revert(); innerErr != nil {
 			return false, fmt.Errorf("when reverting compressor because writing failed: %w\noriginal error: %w", innerErr, err)
```
Comment (ivokub) [inline prover/lib/compressor/blob/v1/blob_maker.go:169]: Can use `errors.Join` for simplifying error handling.

File: prover/lib/compressor/blob/v1/blob_maker.go [hunk 6]
```diff
@@ -191,45 +189,80 @@ func (bm *BlobMaker) Write(rlpBlock []byte, forceReset bool) (ok bool, err error
...
 	}
 
 	// copy the compressed data to the blob
 	bm.packBuffer.Reset()
 	n2, err := encode.PackAlign(&bm.packBuffer, bm.buf.Bytes(), fr381.Bits-1, encode.WithAdditionalInput(bm.compressor.Bytes()))
 	if err != nil {
-		bm.compressor.Revert()
-		bm.Header.removeLastBlock()
+		err = fmt.Errorf("when packing blob: %w", err)
+		innerErr := revert()
```
Comment (ivokub) [inline prover/lib/compressor/blob/v1/blob_maker.go:262]: Can use `errors.Join`

File: prover/lib/compressor/blob/v1/blob_maker.go [hunk 7]
```diff
@@ -191,45 +189,80 @@ func (bm *BlobMaker) Write(rlpBlock []byte, forceReset bool) (ok bool, err error
...
+			err = fmt.Errorf("when recompressing the blob: %w", err)
+
+			if innerErr := revert(); innerErr != nil {
+				err = fmt.Errorf("%w\n\tto recover from write failure: %w", innerErr, err)
+			}
+
+			return false, err
+		}
+		if fitsInBlob() {
+			goto bypass
```
Comment (ivokub) [inline prover/lib/compressor/blob/v1/blob_maker.go:230]: This path doesn't seem to be tested - e.g. first when compressing single block doesn't fit but recompressing full payload then fits.

**Review by ivokub — APPROVED** [review]
Seems good to me. Before merging can you confirm that by referencing to the same buffer when recompressing we don't have some internal overwrites of bytes.Buffer which could cause issues?

Also some comments about simplifying error handling.

That said - imo the current approach may be error prone when wanting to make changes in the future. It heavily relies on reverting the compressor in case the block doesnt fit (or when we only run CanWrite). And the state logic imo is complex enough that it is diffcult to follow and test.

What I would recommend is to add a method to clone the compressor using a pool and then perform operations on the cloned compressor. If write succeeds then we update the reference in `BlobMaker`, otherwise we discard the cloned compressor (by resetting and putting it back into the pool). Internally it would imo also require using a pool of `bytes.Buffer`.

Imo trying to optimize for memory here doesn't make a lot of sense as the blocks/blobs in general are small (imo blob maximally 2MB and maximum uncompressed data assuming 5x compression 10MB) relative to the rest of memory usage.

From my side this is not blocker for merging the PR for now, but would definitely simplify the implementation and guard against bugs in the future.

**Pull request merged**